sdk/core/azure-core-http-okhttp/src/main/java/com/azure/core/http/okhttp/implementation/OkHttpFluxRequestBody.java

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.http.okhttp.implementation;
+
+import com.azure.core.http.HttpHeaders;
+import com.azure.core.implementation.util.BinaryDataContent;
+import com.azure.core.util.logging.ClientLogger;
+import okhttp3.MediaType;
+import okio.BufferedSink;
+import reactor.core.Exceptions;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.time.Duration;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * An {@link okhttp3.RequestBody} subtype that sends {@link BinaryDataContent}
+ * as {@link Flux} of {@link ByteBuffer} in an unbuffered manner.
+ * This class accepts any {@link BinaryDataContent} as catch-all for backwards compatibility
+ * but ideally should be used only with reactive payloads.
+ */
+public class OkHttpFluxRequestBody extends OkHttpStreamableRequestBody<BinaryDataContent> {
+
+    private static final ClientLogger LOGGER = new ClientLogger(OkHttpFluxRequestBody.class);
+
+    private final AtomicBoolean bodySent = new AtomicBoolean(false);
+    private final Duration writeTimeout;
+
+    public OkHttpFluxRequestBody(
+        BinaryDataContent content, HttpHeaders httpHeaders, MediaType mediaType, Duration writeTimeout) {
+        super(content, httpHeaders, mediaType);
+        this.writeTimeout = writeTimeout;
+    }
+
+    @Override
+    public void writeTo(BufferedSink bufferedSink) throws IOException {
+        if (bodySent.compareAndSet(false, true)) {
+            Mono<BufferedSink> requestSendMono = content.toFluxByteBuffer()
+                .publishOn(Schedulers.boundedElastic())
+                .reduce(bufferedSink, (sink, buffer) -> {
+                    try {
+                        while (buffer.hasRemaining()) {
+                            sink.write(buffer);
+                        }
+                        return sink;
+                    } catch (IOException e) {
+                        throw Exceptions.propagate(e);
+                    }
+                });
+
+            // The blocking happens on OkHttp thread pool.
+            if (writeTimeout != null) {
+                requestSendMono.block(writeTimeout);
```

Comment:
Is this timeout in the right place? I'd have expected the write timeout to behave closer to the network layer or in the `sink.write` above. Maybe

```java
Mono<BufferedSink> requestSendMono = content.toFluxByteBuffer()
    .publishOn(Schedulers.boundedElastic())
    .timeout(writeTimeout)
    .reduce(....);
```

Reply:
This javadoc to me sounds like write timeout applies to whole outbound operation.

https://github.com/Azure/azure-sdk-for-java/blob/afde57efa23a2232d90319d1125a91b765622dc7/sdk/core/azure-core-http-okhttp/src/main/java/com/azure/core/http/okhttp/OkHttpAsyncHttpClientBuilder.java#L123-L139

Reply:
this is getting replaced with `callTimeout`. The write timeout is behaving properly and is terminating this whenever socket is not fast enough.

The OkHttp client docs have better wording for write timeout.